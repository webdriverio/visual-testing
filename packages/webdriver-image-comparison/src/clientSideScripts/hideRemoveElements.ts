/**
 * Hide or remove elements on the page
 */
export default function hideRemoveElements(
    hideRemoveElements: {
        hide: (HTMLElement | HTMLElement[])[];
        remove: (HTMLElement | HTMLElement[])[];
    },
    hideRemove: boolean,
) {
    const visitedSelectors: Record<string, boolean> = {}
    hideRemoveElements.hide.forEach((element) => {
        if (Array.isArray(element)) {
            return element.forEach((singleElement: HTMLElement) => hideRemoveEl(singleElement, 'visibility', hideRemove))
        }
        hideRemoveEl(element, 'visibility', hideRemove)
    })

    hideRemoveElements.remove.forEach((element) => {
        if (Array.isArray(element)) {
            return element.forEach((singleElement: HTMLElement) => hideRemoveEl(singleElement, 'display', hideRemove))
        }
        hideRemoveEl(element, 'display', hideRemove)
    })

    function hideRemoveEl(el: HTMLElement, prop: string, hideRemove: boolean) {
        // @ts-ignore
        if (el.style) {
            // Here we get the HTMLElement
            // @ts-ignore
            setPropertyToElement(el, prop, hideRemove)
        } else {
            // Here we have the WebElement, with the web element we can have 2 types of selectors
            // css and xpath, transform them into HTML
            // This is an anti pattern, but I don't know how to do this better with XPATH selection
            try {
                // @ts-ignore
                const selector = el.selector

                if (!visitedSelectors[selector]) {
                    visitedSelectors[selector] = true
                    const elems = document.querySelectorAll(selector)

                    elems.forEach((singleEl) => setPropertyToElement(singleEl, prop, hideRemove))
                }
            } catch (_e) {
                // 99.99% sure that we have XPATH here
                // @ts-ignore
                return getElementsByXpath(el.selector).forEach((singleEl) => setPropertyToElement(singleEl, prop, hideRemove))
            }
        }
    }

    function setPropertyToElement(el: HTMLElement, prop: string, hideRemove: boolean) {
        const value = prop === 'visibility' ? 'hidden' : 'none'
        // @ts-ignore
        el.style[prop] = hideRemove ? value : ''

        hideRemove ? el.style.setProperty(prop, value, 'important') : el.style.removeProperty(prop)
    }

    // Stupid TypeScript =)
    function getElementsByXpath(xpathToExecute: string): any[] {
        const result = []
        const nodesSnapshot = document.evaluate(xpathToExecute, document, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null)

        for (let i = 0; i < nodesSnapshot.snapshotLength; i++) {
            result.push(nodesSnapshot.snapshotItem(i))
        }

        return result
    }
}
