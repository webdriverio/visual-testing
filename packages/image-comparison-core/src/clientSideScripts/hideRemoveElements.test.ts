// @vitest-environment jsdom

import { describe, it, expect } from 'vitest'
import hideRemoveElements from './hideRemoveElements.js'

describe('hideRemoveElements', () => {
    it('should be able to hide elements and put them back again', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span id="id-1">Hello</span>' +
            '   <span id="id-2">Hello</span>' +
            '   <div>' +
            '     <span id="id-3">Hello</span>' +
            '     <span id="id-4">Hello</span>' +
            '  </div>' +
            '</div>'

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [<HTMLElement>document.querySelector('#id-1'), <HTMLElement>document.querySelector('#id-3')],
                remove: [],
            },
            true,
        )

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [<HTMLElement>document.querySelector('#id-1'), <HTMLElement>document.querySelector('#id-3')],
                remove: [],
            },
            false,
        )

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
    })

    it('should be able to hide elements and put them back again when an array of hidden elements is provided', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span class="hide">Hello</span>' +
            '   <span class="hide">Hello</span>' +
            '   <div>' +
            '     <span id="id-3">Hello</span>' +
            '     <span class="hide">Hello</span>' +
            '  </div>' +
            '</div>'

        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[0].style.visibility).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[1].style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[2].style.visibility).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [[...(<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))]],
                remove: [],
            },
            true,
        )

        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[0].style.visibility).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[1].style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[2].style.visibility).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [[...(<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))]],
                remove: [],
            },
            false,
        )

        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[0].style.visibility).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[1].style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.hide')))[2].style.visibility).toMatchSnapshot()
    })

    it('should be able to remove elements and put them back again', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span id="id-1">Hello</span>' +
            '   <span id="id-2">Hello</span>' +
            '   <div>' +
            '     <span id="id-3">Hello</span>' +
            '     <span id="id-4">Hello</span>' +
            '  </div>' +
            '</div>'

        expect((<HTMLElement>document.querySelector('#id-2')).style.display).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-4')).style.display).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [],
                remove: [<HTMLElement>document.querySelector('#id-2'), <HTMLElement>document.querySelector('#id-4')],
            },
            true,
        )

        expect((<HTMLElement>document.querySelector('#id-2')).style.display).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-4')).style.display).toMatchSnapshot()

        hideRemoveElements(
            {
                remove: [<HTMLElement>document.querySelector('#id-2'), <HTMLElement>document.querySelector('#id-4')],
                hide: [],
            },
            false,
        )

        expect((<HTMLElement>document.querySelector('#id-2')).style.display).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-4')).style.display).toMatchSnapshot()
    })

    it('should be able to remove elements and put them back again when an array of to be removed elements is provided', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span class="remove">Hello</span>' +
            '   <span class="remove">Hello</span>' +
            '   <div>' +
            '     <span id="id-3">Hello</span>' +
            '     <span class="remove">Hello</span>' +
            '  </div>' +
            '</div>'

        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[0].style.display).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[1].style.display).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.display).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[2].style.display).toMatchSnapshot()

        hideRemoveElements(
            {
                remove: [[...(<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))]],
                hide: [],
            },
            true,
        )

        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[0].style.display).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[1].style.display).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.display).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[2].style.display).toMatchSnapshot()

        hideRemoveElements(
            {
                remove: [[...(<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))]],
                hide: [],
            },
            false,
        )

        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[0].style.display).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[1].style.display).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.display).toMatchSnapshot()
        expect((<HTMLElement[]>(<unknown>document.querySelectorAll('.remove')))[2].style.display).toMatchSnapshot()
    })

    it('should be able to find and hide single element based on xpath', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span id="id-1">Hello</span>' +
            '   <span id="id-2">Hello</span>' +
            '   <div>' +
            '     <span id="id-3">Hello</span>' +
            '     <span id="id-4">Hello</span>' +
            '  </div>' +
            '</div>'

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [<any>{ selector: "//span[@id='id-1']" }, <any>{ selector: "//span[@id='id-3']" }],
                remove: [],
            },
            true,
        )

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
    })

    it('should be able to find and hide elements based on xpath', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span id="id-1">Hello</span>' +
            '   <span id="id-2">Hello</span>' +
            '   <div>' +
            '     <span id="id-3">Hello</span>' +
            '     <span id="id-4">Hello</span>' +
            '  </div>' +
            '</div>'

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-2')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-4')).style.visibility).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [[<any>{ selector: '//span' }]],
                remove: [],
            },
            true,
        )

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-2')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-4')).style.visibility).toMatchSnapshot()
    })

    it('should be able to find and hide a single element based on a css selector', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span id="id-1">Hello</span>' +
            '   <span id="id-2">Hello</span>' +
            '   <div>' +
            '     <span id="id-3">Hello</span>' +
            '     <span class="hide">Hello</span>' +
            '  </div>' +
            '</div>'

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-2')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('.hide')).style.visibility).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [<any>{ selector: '.hide' }],
                remove: [],
            },
            true,
        )

        expect((<HTMLElement>document.querySelector('#id-1')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-2')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('#id-3')).style.visibility).toMatchSnapshot()
        expect((<HTMLElement>document.querySelector('.hide')).style.visibility).toMatchSnapshot()
    })

    it('should be able to find and hide elements based on a css selector', () => {
        document.body.innerHTML =
            '<div>' +
            '   <span class="hide">Hello</span>' +
            '   <span class="hide">Hello</span>' +
            '   <div>' +
            '     <span class="hide">Hello</span>' +
            '     <span class="hide">Hello</span>' +
            '  </div>' +
            '</div>'

        expect(<HTMLElement[]>(<unknown>document.querySelectorAll('.hide'))).toMatchSnapshot()

        hideRemoveElements(
            {
                hide: [[<any>{ selector: '.hide' }, <any>{ selector: '.hide' }]],
                remove: [],
            },
            true,
        )

        expect(<HTMLElement[]>(<unknown>document.querySelectorAll('.hide'))).toMatchSnapshot()
    })
})
