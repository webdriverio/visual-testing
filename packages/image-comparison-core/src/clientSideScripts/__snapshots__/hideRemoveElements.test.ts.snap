// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 1`] = `""`;

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 2`] = `""`;

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 3`] = `""`;

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 4`] = `""`;

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 5`] = `""`;

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 6`] = `""`;

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 7`] = `""`;

exports[`hideRemoveElements > should be able to find and hide a single element based on a css selector 8`] = `"hidden"`;

exports[`hideRemoveElements > should be able to find and hide elements based on a css selector 1`] = `
NodeList [
  <span
    class="hide"
  >
    Hello
  </span>,
  <span
    class="hide"
  >
    Hello
  </span>,
  <span
    class="hide"
  >
    Hello
  </span>,
  <span
    class="hide"
  >
    Hello
  </span>,
]
`;

exports[`hideRemoveElements > should be able to find and hide elements based on a css selector 2`] = `
NodeList [
  <span
    class="hide"
    style="visibility: hidden;"
  >
    Hello
  </span>,
  <span
    class="hide"
    style="visibility: hidden;"
  >
    Hello
  </span>,
  <span
    class="hide"
    style="visibility: hidden;"
  >
    Hello
  </span>,
  <span
    class="hide"
    style="visibility: hidden;"
  >
    Hello
  </span>,
]
`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 1`] = `""`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 2`] = `""`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 3`] = `""`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 4`] = `""`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 5`] = `"hidden"`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 6`] = `"hidden"`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 7`] = `"hidden"`;

exports[`hideRemoveElements > should be able to find and hide elements based on xpath 8`] = `"hidden"`;

exports[`hideRemoveElements > should be able to find and hide single element based on xpath 1`] = `""`;

exports[`hideRemoveElements > should be able to find and hide single element based on xpath 2`] = `""`;

exports[`hideRemoveElements > should be able to find and hide single element based on xpath 3`] = `"hidden"`;

exports[`hideRemoveElements > should be able to find and hide single element based on xpath 4`] = `"hidden"`;

exports[`hideRemoveElements > should be able to hide elements and put them back again 1`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again 2`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again 3`] = `"hidden"`;

exports[`hideRemoveElements > should be able to hide elements and put them back again 4`] = `"hidden"`;

exports[`hideRemoveElements > should be able to hide elements and put them back again 5`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again 6`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 1`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 2`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 3`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 4`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 5`] = `"hidden"`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 6`] = `"hidden"`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 7`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 8`] = `"hidden"`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 9`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 10`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 11`] = `""`;

exports[`hideRemoveElements > should be able to hide elements and put them back again when an array of hidden elements is provided 12`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again 1`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again 2`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again 3`] = `"none"`;

exports[`hideRemoveElements > should be able to remove elements and put them back again 4`] = `"none"`;

exports[`hideRemoveElements > should be able to remove elements and put them back again 5`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again 6`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 1`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 2`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 3`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 4`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 5`] = `"none"`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 6`] = `"none"`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 7`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 8`] = `"none"`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 9`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 10`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 11`] = `""`;

exports[`hideRemoveElements > should be able to remove elements and put them back again when an array of to be removed elements is provided 12`] = `""`;
